In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

# Overview of the Data

In [2]:
%%sql
SELECT *
FROM sqlite_master
WHERE type = 'table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

In [3]:
%%sql
SELECT *
FROM facts
LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

# Summary Statistics (looking for any outliers)

In [4]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

We see a few interesting things in the summary statistics:
* There's a country with a population of 0
* There's a country with a population of 7256490011 (or more than 7.2 billion people)

# Exploring Outliers

### _countries with the minimum population_

In [5]:
%%sql
SELECT MIN(population)
FROM facts;

 * sqlite:///factbook.db
Done.


[(0,)]

In [6]:
%%sql
SELECT name AS Country
FROM facts
WHERE population = (SELECT MIN(population)
                    FROM facts
                    );

 * sqlite:///factbook.db
Done.


[('Antarctica',)]

### _countries with the maximum population_

In [7]:
%%sql
SELECT MAX(population)
FROM facts;

 * sqlite:///factbook.db
Done.


[(7256490011,)]

In [8]:
%%sql
SELECT name AS Country
FROM facts
WHERE population = (SELECT MAX(population)
                    FROM facts
                    );

 * sqlite:///factbook.db
Done.


[('World',)]

* It seems like the table contains a row for the whole world, which explains the population of over 7.2 billion. 
* It also seems like the table contains a row for Antarctica, which explains the population of 0. 

* Now that we know this, we should recalculate the summary statistics we calculated earlier — this time excluding the row for the whole world.

# Exploring Average Population and Area

In [9]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               );

 * sqlite:///factbook.db
Done.


[(0, 1367485388, 0.0, 4.02)]

In [10]:
%%sql
SELECT AVG(population) AS average_population
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               );

 * sqlite:///factbook.db
Done.


[(32242666.56846473,)]

In [11]:
%%sql
SELECT AVG(area) AS average_area
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               );

 * sqlite:///factbook.db
Done.


[(555093.546184739,)]

# Finding Densely Populated Countries

Densely populated countries are defined as :
* Above-average values for population.
* Below-average values for area.

In [12]:
%%sql
SELECT name AS country, population, area, ROUND(population / CAST(area AS FLOAT),1) AS population_area_ratio
FROM facts
WHERE population > (SELECT AVG(population) AS average_population
                    FROM facts
                    WHERE name != (SELECT name AS Country
                                   FROM facts
                                   WHERE population = (SELECT MAX(population)
                                                       FROM facts
                                                       )
                                   )                    
                    )
AND area < (SELECT AVG(area) AS average_area
            FROM facts
            WHERE name != (SELECT name AS Country
                           FROM facts
                           WHERE population = (SELECT MAX(population)
                                               FROM facts
                                               )
                           )            
            )
ORDER BY population_area_ratio DESC;

 * sqlite:///factbook.db
Done.


[('Bangladesh', 168957745, 148460, 1138.1),
 ('Korea, South', 49115196, 99720, 492.5),
 ('Philippines', 100998376, 300000, 336.7),
 ('Japan', 126919659, 377915, 335.8),
 ('Vietnam', 94348835, 331210, 284.9),
 ('United Kingdom', 64088222, 243610, 263.1),
 ('Germany', 80854408, 357022, 226.5),
 ('Italy', 61855120, 301340, 205.3),
 ('Uganda', 37101745, 241038, 153.9),
 ('Thailand', 67976405, 513120, 132.5),
 ('Poland', 38562189, 312685, 123.3),
 ('Spain', 48146134, 505370, 95.3),
 ('Iraq', 37056169, 438317, 84.5),
 ('Morocco', 33322699, 446550, 74.6)]

# Country with most people

In [13]:
%%sql
SELECT name AS most_populated_country, population
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
ORDER BY population DESC
LIMIT 1;

 * sqlite:///factbook.db
Done.


[('China', 1367485388)]

# Country with the highest growth rate

In [14]:
%%sql
SELECT name AS highest_growth_rate_country, population_growth
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
ORDER BY population_growth DESC
LIMIT 1;

 * sqlite:///factbook.db
Done.


[('South Sudan', 4.02)]

# Countries having the highest ratios of water to land

In [15]:
%%sql
SELECT name AS country, CAST(area_water AS FLOAT) / area_land AS water_to_land_ratio
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
ORDER BY water_to_land_ratio DESC;

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory', 905.6666666666666),
 ('Virgin Islands', 4.520231213872832),
 ('Puerto Rico', 0.5547914317925592),
 ('Bahamas, The', 0.3866133866133866),
 ('Guinea-Bissau', 0.2846728307254623),
 ('Malawi', 0.25939625850340137),
 ('Netherlands', 0.22571032366565366),
 ('Uganda', 0.22292237442922375),
 ('Eritrea', 0.16435643564356436),
 ('Liberia', 0.15623961794019933),
 ('Bangladesh', 0.1405085657217485),
 ('Gambia, The', 0.116600790513834),
 ('Taiwan', 0.1153130812151271),
 ('Finland', 0.11299639583299047),
 ('India', 0.10563390940312317),
 ('Canada', 0.09799992456155804),
 ('Sweden', 0.09738384490720996),
 ('Colombia', 0.09647636468662751),
 ('Brunei', 0.0949667616334283),
 ('Guyana', 0.09205025171578214),
 ('French Polynesia', 0.08884243532793311),
 ('Nicaragua', 0.08650720893407784),
 ('Burundi', 0.0837227414330218),
 ('Iran', 0.07612978626856316),
 ('United States', 0.07255091319919764),
 ('Tanzania', 0.06942876495822985),
 ('Vietnam', 0.06817815332021801),
 ('Rwanda', 0.06769904329495703),
 ('Estonia', 0.06700009436633009),
 ('Norway', 0.06415101780585115),
 ('Armenia', 0.054604120129064286),
 ('Tunisia', 0.05310247167868177),
 ('Uzbekistan', 0.051716031969910675),
 ('Indonesia', 0.0513367141963679),
 ('Suriname', 0.05012820512820513),
 ('Albania', 0.04927366961092051),
 ('Ghana', 0.048344635723169826),
 ('Azerbaijan', 0.048058187803313605),
 ('Togo', 0.044129815206398824),
 ('Russia', 0.04399263341674328),
 ('Kyrgyzstan', 0.04249195781043894),
 ('Tonga', 0.04184100418410042),
 ('Ukraine', 0.04180691488443547),
 ('Lithuania', 0.041799617102744095),
 ('West Bank', 0.03900709219858156),
 ('Gabon', 0.03880978161735884),
 ('Turkmenistan', 0.038665333134722195),
 ('Hungary', 0.038166235157575217),
 ('Latvia', 0.03759096531671192),
 ('Romania', 0.03697404422095689),
 ('Japan', 0.036846509458551106),
 ('Burma', 0.03530178666519767),
 ('Congo, Democratic Republic of the', 0.03432216697661452),
 ('Venezuela', 0.034011677342554274),
 ('Pakistan', 0.03271606940165397),
 ('Hong Kong', 0.032618825722273995),
 ('Solomon Islands', 0.03251625812906453),
 ('Switzerland', 0.032002400180013504),
 ('Moldova', 0.029187315679061143),
 ('China', 0.029009018475490568),
 ('Korea, South', 0.028889806025588115),
 ('Poland', 0.027707022070302872),
 ('Iceland', 0.02743142144638404),
 ('Montenegro', 0.026761819803746655),
 ('Nepal', 0.02671763712844696),
 ('Botswana', 0.026467630088401885),
 ('Laos', 0.025996533795493933),
 ('Cambodia', 0.025606888932951873),
 ('Italy', 0.024478139661385734),
 ('Ecuador', 0.02427386116940771),
 ('Germany', 0.023948008443465493),
 ('Paraguay', 0.023785432743857318),
 ('Belarus', 0.023164120256283883),
 ('Papua New Guinea', 0.022037715850373184),
 ('Bulgaria', 0.022029883213966393),
 ('Senegal', 0.02177323014595128),
 ('Israel', 0.02164289227742253),
 ('Czech Republic', 0.02097168822090178),
 ('Ireland', 0.02017914434621024),
 ('Kenya', 0.019726253645851635),
 ('Chad', 0.019695044472681066),
 ('Slovakia', 0.01933270969753664),
 ('Brazil', 0.018859459161966657),
 ('Tajikistan', 0.018302593456292843),
 ('Turkey', 0.018099559269884827),
 ('Benin', 0.018079586339064563),
 ('Austria', 0.017296379404451452),
 ('Lebanon', 0.016617790811339198),
 ('Mozambique', 0.016531447900506115),
 ('Chile', 0.01652299236903949),
 ('Saint Lucia', 0.0165016501650165),
 ('Somalia', 0.016450488334021428),
 ('New Caledonia', 0.016415868673050615),
 ('Mali', 0.016392529032363812),
 ('Guatemala', 0.016144234268703515),
 ('Argentina', 0.015971849204696183),
 ('Denmark', 0.015553565537069332),
 ('El Salvador', 0.015443270112446311),
 ('Sri Lanka', 0.01516323688689463),
 ('Jamaica', 0.01477241251961961),
 ('Singapore', 0.01455604075691412),
 ('Panama', 0.014527845036319613),
 ('Nigeria', 0.014273667937389104),
 ('Bolivia', 0.014105036365700761),
 ("Cote d'Ivoire", 0.014025024921148543),
 ('Spain', 0.012806124493967693),
 ('Zambia', 0.012402508481324943),
 ('Croatia', 0.011076571265230286),
 ('Macedonia', 0.011009318601816538),
 ('M

### _countries having more water than land_

In [16]:
%%sql
SELECT name AS country, CAST(area_water AS FLOAT) / area_land AS water_to_land_ratio
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
AND water_to_land_ratio > 1
ORDER BY water_to_land_ratio DESC;

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory', 905.6666666666666),
 ('Virgin Islands', 4.520231213872832)]

# Countries that will be adding the most people to their populations next year

In [17]:
%%sql
SELECT name AS country, ROUND((population_growth / 100) * population, 0) AS population_added_next_year
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
ORDER BY population_added_next_year DESC;

 * sqlite:///factbook.db
Done.


[('India', 15270686.0),
 ('China', 6153684.0),
 ('Nigeria', 4448270.0),
 ('Pakistan', 2906653.0),
 ('Ethiopia', 2874562.0),
 ('Bangladesh', 2703324.0),
 ('United States', 2506677.0),
 ('Indonesia', 2355142.0),
 ('Congo, Democratic Republic of the', 1944691.0),
 ('Philippines', 1626074.0),
 ('Egypt', 1583924.0),
 ('Brazil', 1572801.0),
 ('Mexico', 1436494.0),
 ('Tanzania', 1424180.0),
 ('European Union', 1284874.0),
 ('Uganda', 1202097.0),
 ('Iraq', 1085746.0),
 ('Turkey', 1000620.0),
 ('Iran', 981891.0),
 ('Vietnam', 915184.0),
 ('Kenya', 886358.0),
 ('Afghanistan', 755493.0),
 ('Algeria', 727576.0),
 ('South Africa', 713885.0),
 ('Yemen', 660412.0),
 ('Sudan', 621072.0),
 ('Mozambique', 619926.0),
 ('Cameroon', 614846.0),
 ('Madagascar', 614367.0),
 ('Malawi', 596428.0),
 ('Niger', 586486.0),
 ('Ghana', 573943.0),
 ('Burkina Faso', 573630.0),
 ('Burma', 568834.0),
 ('Nepal', 564768.0),
 ('Angola', 545585.0),
 ('Mali', 505275.0),
 ('Colombia', 486062.0),
 ('South Sudan', 484125.0),
 ("Cote d'Ivoire", 444940.0),
 ('Malaysia', 439399.0),
 ('Zambia', 433908.0),
 ('Spain', 428501.0),
 ('Venezuela', 406929.0),
 ('Saudi Arabia', 405184.0),
 ('Argentina', 403917.0),
 ('Burundi', 352347.0),
 ('United Kingdom', 346076.0),
 ('Senegal', 342408.0),
 ('Morocco', 333227.0),
 ('Rwanda', 324140.0),
 ('Zimbabwe', 314473.0),
 ('Guinea', 309818.0),
 ('Peru', 295316.0),
 ('Benin', 290472.0),
 ('France', 286181.0),
 ('Uzbekistan', 271559.0),
 ('Guatemala', 271526.0),
 ('Ukraine', 266577.0),
 ('Canada', 263249.0),
 ('Cambodia', 248198.0),
 ('Australia', 243436.0),
 ('Thailand', 231120.0),
 ('Chad', 219835.0),
 ('Ecuador', 214223.0),
 ('Kazakhstan', 206991.0),
 ('Togo', 203157.0),
 ('Japan', 203071.0),
 ('Somalia', 194280.0),
 ('Sri Lanka', 185249.0),
 ('Bolivia', 168494.0),
 ('Italy', 167009.0),
 ('United Arab Emirates', 149118.0),
 ('Honduras', 146944.0),
 ('Eritrea', 146873.0),
 ('Chile', 143568.0),
 ('Libya', 142983.0),
 ('Tajikistan', 140082.0),
 ('Sierra Leone', 138159.0),
 ('Germany', 137452.0),
 ('Korea, North', 132411.0),
 ('Dominican Republic', 128889.0),
 ('Israel', 125569.0),
 ('Papua New Guinea', 118769.0),
 ('Haiti', 118287.0),
 ('Central African Republic', 114840.0),
 ('Singapore', 107248.0),
 ('Laos', 107129.0),
 ('Liberia', 103633.0),
 ('Tunisia', 98231.0),
 ('Congo, Republic of the', 95102.0),
 ('Azerbaijan', 93895.0),
 ('Belgium', 86062.0),
 ('Mauritania', 80206.0),
 ('Paraguay', 78686.0),
 ('Sweden', 78413.0),
 ('Netherlands', 69486.0),
 ('Korea, South', 68761.0),
 ('Oman', 68040.0),
 ('Qatar', 67381.0),
 ('Jordan', 67376.0),
 ('Romania', 64999.0),
 ('Kyrgyzstan', 62881.0),
 ('Ireland', 61154.0),
 ('Turkmenistan', 59638.0),
 ('Nicaragua', 59079.0),
 ('Norway', 58847.0),
 ('Costa Rica', 58733.0),
 ('Switzerland', 57665.0),
 ('Russia', 56970.0),
 ('West Bank', 54315.0),
 ('Taiwan', 53855.0),
 ('Lebanon', 53188.0),
 ('Gaza Strip', 52520.0),
 ('Panama', 48273.0),
 ('Austria', 47661.0),
 ('Kuwait', 45174.0),
 ('Gambia, The', 42503.0),
 ('Bulgaria', 41684.0),
 ('Mongolia', 39207.0),
 ('Moldova', 36533.0),
 ('New Zealand', 36395.0),
 ('Poland', 34706.0),
 ('Serbia', 33013.0),
 ('Guinea-Bissau', 32970.0),
 ('Gabon', 32913.0),
 ('Bahrain', 32453.0),
 ('Lithuania', 29998.0),
 ('Timor-Leste', 29793.0),
 ('Syria', 27304.0),
 ('Hong Kong', 27136.0),
 ('Botswana', 26411.0),
 ('Finland', 21908.0),
 ('Hungary', 21775.0),
 ('Puerto Rico', 21590.0),
 ('Latvia', 21059.0),
 ('Jamaica', 20061.0),
 ('Belarus', 19179.0),
 ('Equatorial Guinea', 18593.0),
 ('Djibouti', 18223.0),
 ('Czech Republic', 17032.0),
 ('Cyprus', 17006.0),
 ('Cuba', 16547.0),
 ('Western Sahara', 16098.0),
 ('Swaziland', 15935.0),
 ('El Salvador', 15353.0),
 ('Comoros', 13823.0),
 ('Namibia', 13053.0),
 ('Solomon Islands', 12574.0),
 ('Denmark', 12279.0),
 ('Luxembourg', 12146.0),
 ('Portugal', 9743.0),
 ('Albania', 9088.0),
 ('Uruguay', 9023.0),
 ('Mauritius', 8575.0),
 ('Bhutan', 8235.0),
 ('Cabo Verde', 7426.0),
 ('Brunei', 6960.0),
 ('Estonia', 6960.0),
 ('Belize', 6496.0),
 ('

# Countries having a higher death rate than birth rate

In [18]:
%%sql
SELECT name AS country, death_rate, birth_rate,  CAST(death_rate AS FLOAT) / birth_rate AS death_to_birth_ratio
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
AND death_to_birth_ratio > 1;

 * sqlite:///factbook.db
Done.


[('Austria', 9.42, 9.41, 1.0010626992561105),
 ('Belarus', 13.36, 10.7, 1.2485981308411216),
 ('Bosnia and Herzegovina', 9.75, 8.87, 1.0992108229988726),
 ('Bulgaria', 14.44, 8.92, 1.6188340807174888),
 ('Croatia', 12.18, 9.45, 1.288888888888889),
 ('Czech Republic', 10.34, 9.63, 1.0737279335410175),
 ('Estonia', 12.4, 10.51, 1.1798287345385348),
 ('Germany', 11.42, 8.47, 1.3482880755608027),
 ('Greece', 11.09, 8.66, 1.2806004618937643),
 ('Hungary', 12.73, 9.16, 1.3897379912663756),
 ('Italy', 10.19, 8.74, 1.165903890160183),
 ('Japan', 9.51, 7.93, 1.1992433795712485),
 ('Latvia', 14.31, 10.0, 1.431),
 ('Lithuania', 14.27, 10.1, 1.4128712871287128),
 ('Moldova', 12.59, 12.0, 1.0491666666666666),
 ('Monaco', 9.24, 6.65, 1.3894736842105262),
 ('Poland', 10.19, 9.74, 1.0462012320328542),
 ('Portugal', 11.02, 9.27, 1.1887810140237325),
 ('Romania', 11.9, 9.14, 1.3019693654266957),
 ('Russia', 13.69, 11.6, 1.1801724137931033),
 ('Serbia', 13.66, 9.08, 1.5044052863436124),
 ('Slovenia', 11.37, 8.42, 1.350356294536817),
 ('Ukraine', 14.46, 10.72, 1.3488805970149254),
 ('Saint Pierre and Miquelon', 9.72, 7.42, 1.3099730458221026)]

# Countries having the highest population/area ratio

In [19]:
%%sql
SELECT name AS country, population, area, ROUND(population / CAST(area AS FLOAT),1) AS population_area_ratio
FROM facts
WHERE name != (SELECT name AS Country
               FROM facts
               WHERE population = (SELECT MAX(population)
                                   FROM facts
                                   )
               )
ORDER BY population_area_ratio DESC;

 * sqlite:///factbook.db
Done.


[('Macau', 592731, 28, 21169.0),
 ('Monaco', 30535, 2, 15267.5),
 ('Singapore', 5674472, 697, 8141.3),
 ('Hong Kong', 7141106, 1108, 6445.0),
 ('Gaza Strip', 1869055, 360, 5191.8),
 ('Gibraltar', 29258, 6, 4876.3),
 ('Bahrain', 1346613, 760, 1771.9),
 ('Maldives', 393253, 298, 1319.6),
 ('Malta', 413965, 316, 1310.0),
 ('Bermuda', 70196, 54, 1299.9),
 ('Sint Maarten', 39689, 34, 1167.3),
 ('Bangladesh', 168957745, 148460, 1138.1),
 ('Guernsey', 66080, 78, 847.2),
 ('Jersey', 97294, 116, 838.7),
 ('Barbados', 290604, 430, 675.8),
 ('Mauritius', 1339827, 2040, 656.8),
 ('Taiwan', 23415126, 35980, 650.8),
 ('Aruba', 112162, 180, 623.1),
 ('Lebanon', 6184701, 10400, 594.7),
 ('Saint Martin', 31754, 54, 588.0),
 ('San Marino', 33020, 61, 541.3),
 ('Korea, South', 49115196, 99720, 492.5),
 ('Rwanda', 12661733, 26338, 480.7),
 ('West Bank', 2785366, 5860, 475.3),
 ('Nauru', 9540, 21, 454.3),
 ('Tuvalu', 10869, 26, 418.0),
 ('Netherlands', 16947904, 41543, 408.0),
 ('Marshall Islands', 72191, 181, 398.8),
 ('Israel', 8049314, 20770, 387.5),
 ('Burundi', 10742276, 27830, 386.0),
 ('India', 1251695584, 3287263, 380.8),
 ('Belgium', 11323973, 30528, 370.9),
 ('Haiti', 10110019, 27750, 364.3),
 ('Comoros', 780971, 2235, 349.4),
 ('Philippines', 100998376, 300000, 336.7),
 ('Sri Lanka', 22053488, 65610, 336.1),
 ('Japan', 126919659, 377915, 335.8),
 ('Curacao', 148406, 444, 334.2),
 ('Grenada', 110694, 344, 321.8),
 ('Guam', 161785, 544, 297.4),
 ('El Salvador', 6141350, 21041, 291.9),
 ('Vietnam', 94348835, 331210, 284.9),
 ('American Samoa', 54343, 199, 273.1),
 ('Jamaica', 2950210, 10991, 268.4),
 ('Saint Lucia', 163922, 616, 266.1),
 ('Saint Vincent and the Grenadines', 102627, 389, 263.8),
 ('United Kingdom', 64088222, 243610, 263.1),
 ('Puerto Rico', 3598357, 13791, 260.9),
 ('Pakistan', 199085847, 796095, 250.1),
 ('Trinidad and Tobago', 1222363, 5128, 238.4),
 ('Liechtenstein', 37624, 160, 235.2),
 ('Germany', 80854408, 357022, 226.5),
 ('British Virgin Islands', 33454, 151, 221.5),
 ('Luxembourg', 570252, 2586, 220.5),
 ('Dominican Republic', 10478756, 48670, 215.3),
 ('Nepal', 31551305, 147181, 214.4),
 ('Cayman Islands', 56092, 264, 212.5),
 ('Antigua and Barbuda', 92436, 442, 209.1),
 ('Korea, North', 24983205, 120538, 207.3),
 ('Italy', 61855120, 301340, 205.3),
 ('Seychelles', 92430, 455, 203.1),
 ('Sao Tome and Principe', 194006, 964, 201.3),
 ('Saint Kitts and Nevis', 51936, 261, 199.0),
 ('Switzerland', 8121830, 41277, 196.8),
 ('Nigeria', 181562056, 923768, 196.5),
 ('Qatar', 2194817, 11586, 189.4),
 ('Andorra', 85580, 468, 182.9),
 ('Anguilla', 16418, 91, 180.4),
 ('Gambia, The', 1967709, 11300, 174.1),
 ('Kosovo', 1870981, 10887, 171.9),
 ('Kuwait', 2788534, 17818, 156.5),
 ('Uganda', 37101745, 241038, 153.9),
 ('Isle of Man', 87545, 572, 153.1),
 ('Malawi', 17964697, 118484, 151.6),
 ('Micronesia, Federated States of', 105216, 702, 149.9),
 ('Tonga', 106501, 747, 142.6),
 ('China', 1367485388, 9596960, 142.5),
 ('Guatemala', 14918999, 108889, 137.0),
 ('Cabo Verde', 545993, 4033, 135.4),
 ('Czech Republic', 10644842, 78867, 135.0),
 ('Indonesia', 255993674, 1904569, 134.4),
 ('Togo', 7552318, 56785, 133.0),
 ('Thailand', 67976405, 513120, 132.5),
 ('Kiribati', 105711, 811, 130.3),
 ('Denmark', 5581503, 43094, 129.5),
 ('Cyprus', 1189197, 9251, 128.5),
 ('Akrotiri', 15700, 123, 127.6),
 ('Poland', 38562189, 312685, 123.3),
 ('Dhekelia', 15700, 130, 120.8),
 ('European Union', 513949445, 4324782, 118.8),
 ('Portugal', 10825309, 92090, 117.6),
 ('Azerbaijan', 9780780, 86600, 112.9),
 ('Northern Mariana Islands', 52344, 464, 112.8),
 ('Tokelau', 1337, 12, 111.4),
 ('Slovakia', 5445027, 49035, 111.0),
 ('Ghana', 26327649, 238533, 110.4),
 ('Wallis and Futuna', 15613, 142, 110.0),
 ('Hungary', 9897541, 93028, 106.4),
 ('Albania', 3029278, 28748, 105.4),
 ('Moldova', 3546847, 33851, 104.8),
 ('France', 66553766, 643801, 103.4),
 ('Austria', 8665550, 83871, 103.3),
 ('Armenia', 3056382, 29743, 102.8),
 ('Turkey', 79414269, 783

In [20]:
%%sql
SELECT AVG(population) AS avg_pop_top_20, AVG(area) AS avg_area_top_20, 
       (SELECT population
        FROM facts
        WHERE name = 'Bangladesh'
        ) AS bangladesh_population,
       (SELECT area
        FROM facts
        WHERE name = 'Bangladesh'
        ) AS bangladesh_area
FROM (SELECT name AS country, population, area, ROUND(population / CAST(area AS FLOAT),1) AS population_area_ratio
      FROM facts
      WHERE name != (SELECT name AS Country
                      FROM facts
                      WHERE population = (SELECT MAX(population)
                                          FROM facts
                                          )
                      )
      ORDER BY population_area_ratio DESC
      LIMIT 20
      );

 * sqlite:///factbook.db
Done.


[(10904808.3, 10070.05, 168957745, 148460)]

In [21]:
%%sql
SELECT bangladesh_population / avg_pop_top_20 AS population_comparison,
       bangladesh_area / avg_area_top_20 AS area_comparison
FROM (SELECT AVG(population) AS avg_pop_top_20, AVG(area) AS avg_area_top_20, 
              (SELECT population
              FROM facts
              WHERE name = 'Bangladesh'
              ) AS bangladesh_population,
              (SELECT area
              FROM facts
              WHERE name = 'Bangladesh'
              ) AS bangladesh_area
      FROM (SELECT name AS country, population, area, ROUND(population / CAST(area AS FLOAT),1) AS population_area_ratio
              FROM facts
              WHERE name != (SELECT name AS Country
                              FROM facts
                              WHERE population = (SELECT MAX(population)
                                                  FROM facts
                                                  )
                              )
              ORDER BY population_area_ratio DESC
              LIMIT 20
              )        
      );

 * sqlite:///factbook.db
Done.


[(15.493875761208933, 14.742727195992076)]

* The top 10 countries in the tables for most densely populated countries and highest population/area ratio, we see no match.
  - It is seen that the top countries in the highest population/area ratio table all have very small areas(less than 1200 sq. kms), which massively drives up their population density(even if their populations are less - which is usually the case). Hence overall area of a country seems to be having more influence than overall population numbers.
  - Only Bangladesh seems to be present at the top (i.e. top 20) amongst countries having highest population to area ratio. This is primarily due to the extremely high number of people in the country (approx. 169 million -> approx. 15.5 times the average top 20 population) and despite having a much larger country area (148460 sq kms -> approx. 14.7 times the average area of top areas) than many top countries in the aforementioned table.
    - The ratio of these comparisons is slightly greater than 1, which is why Bangladesh is in the middle of top 20